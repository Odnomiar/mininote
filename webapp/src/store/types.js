const state = {
  SELECTED_NOTE_ID: 'selectedNoteId',
  NOTEBOOKS: 'notebooks',
  VERSION: 'version',
}
const getters = {
  LOADED: 'loaded',
  DIRTY: 'dirty',
  NOTES: 'notes',
  SELECTED_NOTE: 'selectedNote',
  NOTE_BY_ID: 'noteById',
  CURRENT_CONTENT: 'currentContent',
  CURRENT_TITLE: 'currentTitle',
}
const actions = {
  LOAD_NOTEBOOK: 'loadNotebook',
  LIST_NOTEBOOKS: 'listNotebooks',
  CREATE_NOTEBOOK: 'createNotebook',
  UPDATE_NOTEBOOK: 'updateNotebook',
  UPDATE_NOTE: 'updateNote',
  APPLY_CHANGES: 'applyChanges',
}
const mutations = {
  SELECT_NOTE: 'selectNote',
  SET_LOAD_NOTEBOOK: 'setLoadNotebook',
  SELECT_FIRST: 'selectFirst',
  ADD_CHANGE: 'addChange',
  REVERT_CHANGES: 'revertChanges',
  RESET: 'reset',
}

export { state, getters, actions, mutations }
