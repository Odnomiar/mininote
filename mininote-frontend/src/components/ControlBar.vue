<template>
  <div>
    <b-modal id="discardModal" ref="discardModalRef" title="Save changes?" @ok="reset">You're about to discard all recent changes to your notebook. Are you sure you want to proceed?</b-modal>
    <b-modal id="settingsModal" ref="settingsModalRef" size="lg" title="Notebook Settings" @ok="updateNotebook" ok-title="Save">
      <b-container fluid>
        <b-row>
          <b-col sm="3">
            <label for="newNameInput">Name:</label>
          </b-col>
          <b-col>
            <b-form-input id="newNameInput" type="text" placeholder="Enter new name for this notebook." v-model="inputs.newName" />
          </b-col>
        </b-row>
        <b-row style="margin-top: 15px">
          <b-col sm="3">
            <label for="newPasswordInput">Password:</label>
          </b-col>
          <b-col>
            <b-form-input id="newPasswordInput" type="password" placeholder="Enter new password for this notebook." v-model="inputs.newPassword" />
          </b-col>
        </b-row>
      </b-container>
    </b-modal>
    <div class="container-fluid">
      <div class="row">
        <div class="col-4"></div>
        <div class="col-4">
          <div class="input-group" id="notebook-chooser">
            <div class="input-group-btn">
              <button type="button" class="btn btn-danger" @click="close" :disabled="dirty">&#x2573;</button>
            </div>
            <input type="text" class="form-control" placeholder="Open or create notebook ..." ref="refNotebookInput" v-model="inputs.name" v-if="!state.opening && !state.creating" :disabled="state.loaded" @keyup.enter="tryNotebook" autofocus />
            <input type="password" class="form-control" placeholder="Choose a password for the new notebook..." ref="refCreatePasswordInput" v-model="inputs.password" v-if="state.creating" @keyup.enter="createNotebook" />
            <input type="password" class="form-control" placeholder="Enter password ..." ref="refOpenPasswordInput" v-model="inputs.password" v-if="state.opening" @keyup.enter="openNotebook" />
            <div class="input-group-btn">
              <button type="button" class="btn btn-primary" :disabled="!inputs.name || state.loaded" @click="tryNotebook" v-if="!state.opening && !state.creating">Open</button>
              <button type="button" class="btn btn-primary" :disabled="!inputs.password" @click="openNotebook" v-if="state.opening">Open</button>
              <button type="button" class="btn btn-primary" :disabled="!inputs.password" @click="createNotebook" v-if="state.creating">Create</button>
            </div>
          </div>
        </div>
        <div class="col-2"></div>
        <div class="col-2 action-buttons-container">
          <button class="btn btn-primary float-right" @click="saveNotes" v-if="dirty">&#x1f4be;</button>
          <button class="btn btn-primary float-right" @click="tryReset" v-if="dirty">&#x21ba;</button>
          <button class="btn btn-primary float-right" v-if="state.loaded" v-b-modal.settingsModal>&#x2699;</button>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import api from "../api";
import { md5 } from "../lib/md5";
import { mapState, mapGetters } from "vuex";
import { ApiError, UnauthorizedError, NotFoundError } from "../lib/errors";

export default {
  name: "control-bar",
  data() {
    return {
      inputs: {
        name: '',
        password: '',
        newName: '',
        newPassword: ''
      },
      state: {
        opening: false,
        creating: false,
        loaded: false
      }
    };
  },
  computed: {
    ...mapState({
      notebook: state => state.notebook
    }),
    ...mapGetters(["dirty"])
  },
  methods: {
    handleError: function(err) {
      if (err instanceof UnauthorizedError) this.$emit('alert', 'You are not authorized to access this note. Password wrong?')
      else if (err instanceof NotFoundError) this.$emit('alert', 'Resource not found')
      else this.$emit('alert', 'An error has occured. Sorry.')
      this.reset();
    },
    tryReset: function() {
      if (!this.dirty) this.reset();
      else this.$refs.discardModalRef.show();
    },
    reset: function() {
      this.$store.commit('revertChanges');
    },
    close: function() {
      this.inputs = {
        name: '',
        password: '',
        newName: '',
        newPassword: ''
      };
      this.state = {
        opening: false,
        creating: false,
        loaded: false
      };
      this.$store.commit('reset');
      setTimeout(() => this.$refs.refNotebookInput.focus(), 0);
    },
    tryNotebook: function() {
      let vm = this;
      if (!this.inputs.name) return;

      api
        .exists(this.inputs.name.toLowerCase())
        .then(exists => {
          if (exists) {
            vm.state.opening = true;
            setTimeout(() => vm.$refs.refOpenPasswordInput.focus(), 0);
          } else {
            vm.state.creating = true;
            setTimeout(() => vm.$refs.refCreatePasswordInput.focus(), 0);
          }
        })
        .catch(() => vm.handleError(null));
    },
    openNotebook: function() {
      let vm = this;
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store
        .dispatch('loadNotebook', {
          id: this.inputs.name.toLowerCase(),
          password: md5(this.inputs.password)
        })
        .then(res => {
          vm.state.opening = false;
          vm.state.loaded = true;
          vm.$store.commit('selectFirst');

          this.inputs.newName = this.inputs.name;
          this.inputs.newPassword = this.inputs.password;
        })
        .catch(vm.handleError);
    },
    createNotebook: function() {
      let vm = this;
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store
        .dispatch('createNotebook', {
          id: this.inputs.name.toLowerCase(),
          password: md5(this.inputs.password)
        })
        .then(res => {
          vm.state.creating = false;
          vm.state.loaded = true;
        })
        .catch(vm.handleError);
    },
    saveNotes: function() {
      let vm = this;
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store.dispatch('applyChanges').catch(vm.handleError);
    },
    updateNotebook: function() {
      let vm = this
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store.dispatch('updateNotebook', { id: this.inputs.newName.toLowerCase(), password: md5(this.inputs.newPassword) })
        .then(() => {
          this.inputs.name = this.inputs.newName,
          this.inputs.password = this.inputs.newPassword
        })
        .catch(vm.handleError)
    }
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
.action-buttons-container button {
  margin-left: 5px;
}

#notebook-chooser input {
  border: 0;
}
</style>
