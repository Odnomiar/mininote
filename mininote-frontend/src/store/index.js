import Vue from 'vue'
import Vuex from 'vuex'

import api from '../api'

Vue.use(Vuex)

const emptyState = {
    notebook: {},
    notebooks: [],
    changes: {
        add: [],
        delete: [],
        update: {}
    },
    selectedNoteId: -1,
    loadNotebookId: '',
    version: require('../../package.json').version
}

/// GETTERS ///
function loaded(state) {
    return state.notebook.hasOwnProperty('notes')
}

function dirty(state) {
    return !!(state.changes.add.length || state.changes.delete.length || Object.keys(state.changes.update).length)
}

function selectedNote(state) {
    return resolveNotes(state).find(n => n.id === state.selectedNoteId)
}

function noteById(state) {
    return (id) => resolveNotes(state).find(n => n.id === id)
}

function resolveNotes(state) {
    let notes = []
    notes = notes.concat(state.notebook.notes || [])
    notes = notes.concat(state.changes.add)
    notes = notes.filter(n1 => !state.changes.delete.some(n2 => n1.id === n2.id))
    return notes
}

function currentContent(state) {
    let id = state.selectedNoteId
    if (id < 0) return ''
    if (state.changes.update.hasOwnProperty(id)) return state.changes.update[id].content
    return selectedNote(state).content
}

function currentTitle(state) {
    let id = state.selectedNoteId
    if (id < 0) return ''
    if (state.changes.update.hasOwnProperty(id)) return state.changes.update[id].title
    return selectedNote(state).title
}

/// MUTATIONS ///
function selectNote(state, id) {
    if (id < 0) return null
    let notes = resolveNotes(state)
    if (notes.some(n => n.id === id)) {
        Vue.set(state, 'selectedNoteId', id)
    }
}

function selectFirst(state) {
    let notes = resolveNotes(state)
    let id = notes.length ? notes[0].id : -1
    Vue.set(state, 'selectedNoteId', id)
}

/// STORE ///
const store = new Vuex.Store({
    state: JSON.parse(JSON.stringify(emptyState)),
    getters: {
        selectedNote,
        dirty,
        notes: resolveNotes,
        currentContent,
        currentTitle,
        loaded,
        noteById
    },
    mutations: {
        setNotebooks(state, notebookList) {
            Vue.set(state, 'notebooks', notebookList)
        },
        setNotebook(state, { id, password }) {
            Vue.set(state.notebook, 'id', id)
            Vue.set(state.notebook, 'password', password)
        },
        setNotes(state, notes) {
            Vue.set(state.notebook, 'notes', notes)
        },
        addNote(state, note) {
            Vue.set(state.notebook, 'notes', state.notebook.notes.concat([note]))
            if (state.selectedNoteId === note.oldId) selectNote(state, note.id)
        },
        removeNote(state, note) {
            Vue.set(state.notebook, 'notes', state.notebook.notes.filter(n => n.id !== note.id))
            if (state.selectedNoteId === note.id && state.notebook.notes.length) {
                Vue.set(state, 'selectedNoteId', state.notes[0].id)
            }
        },
        updateNote(state, note) {
            let notes = resolveNotes(state)
            let found = notes.find(n => n.id === note.id)
            for (let prop in found) {
                Vue.set(found, prop, note[prop])
            }
        },
        selectNote,
        selectFirst,
        setLoadNotebook(state, id) {
            Vue.set(state, 'loadNotebookId', id)
        },
        addChange(state, { type, payload }) {
            switch (type) {
                case 'add':
                case 'delete':
                    Vue.set(state.changes, type, state.changes[type].concat([payload]))
                    break
                case 'update':
                    let hasChange = state.changes.update.hasOwnProperty(payload.id)
                    let originalNote = (state.notebook.notes || []).find(n => n.id === payload.id)

                    if (originalNote && originalNote.content === payload.content && hasChange) {
                        Vue.delete(state.changes.update, payload.id)
                    } else {
                        Vue.set(state.changes.update, payload.id, payload)
                    }
                    break
            }
        },
        removeChange(state, { type, payload }) {
            switch (type) {
                case 'add':
                case 'delete':
                    let idx = state.changes[type].findIndex(n => n.id === payload.id)
                    Vue.delete(state.changes[type], idx)
                    break
                case 'update':
                    Vue.delete(state.changes.update, payload.id)
            }
        },
        revertChanges(state) {
            Vue.set(state.changes, 'add', [])
            Vue.set(state.changes, 'delete', [])

            Object.keys(state.changes.update).forEach(id => {
                let note = (state.notebook.notes || []).find(n => n.id === parseInt(id))
                let content = note.content

                Vue.set(note, 'content', '')
                Vue.set(note, 'content', content)
            })

            Vue.set(state.changes, 'update', {})
        },
        reset(state) {
            for (let prop in emptyState) {
                Vue.set(state, prop, JSON.parse(JSON.stringify(emptyState[prop])))
            }
        }
    },
    actions: {
        listNotebooks({ commit }) {
            return api.list()
                .then(data => commit('setNotebooks', data))
        },
        loadNotebook({ commit }, { id, password }) {
            return api.getNotes(id, password)
                .then(data => commit('setNotes', data))
                .then(() => commit('setNotebook', { id, password }))
        },
        createNotebook({ commit }, { id, password }) {
            return api.create(id, password)
                .then(data => commit('setNotes', data.notes))
                .then(() => commit('setNotebook', { id, password }))
        },
        updateNotebook({ state, commit }, notebook) {
            let updateNotebook = JSON.parse(JSON.stringify(notebook))
            if (notebook.id === state.notebook.id) delete updateNotebook.id
            return api.update(state.notebook.id, state.notebook.password, updateNotebook)
                .then(() => commit('setNotebook', notebook))
        },
        createNote({ state, commit }, note) {
            return api.addNote(state.notebook.id, state.notebook.password, note)
                .then(data => commit('addNote', { ...note, id: data.id, oldId: note.id }))
        },
        updateNote({ state, commit }, note) {
            return api.updateNote(state.notebook.id, state.notebook.password, note)
                .then(() => commit('updateNote', note))
        },
        deleteNote({ state, commit }, note) {
            return api.deleteNote(state.notebook.id, state.notebook.password, note)
                .then(() => commit('removeNote', note))
        },
        applyChanges({ state, commit, dispatch }) {
            let promises = []

            Object.values(state.changes.update).forEach(n => {
                let newNote = state.changes.add.find(n1 => n1.id === n.id)
                let deletedNote = state.changes.delete.find(n1 => n1.id === n.id)

                if (newNote) {
                    commit('updateNote', n)
                    commit('removeChange', { type: 'update', payload: n })
                }
                else if (deletedNote) commit('removeChange', { type: 'update', payload: n })
                else promises.push(dispatch('updateNote', n).then(() => commit('removeChange', { type: 'update', payload: n })))
            })

            state.changes.add.forEach(n => {
                if (state.changes.delete.some(n1 => n1.id === n.id)) {
                    commit('removeChange', { type: 'add', payload: n })
                    commit('removeChange', { type: 'delete', payload: n })
                }
                else promises.push(dispatch('createNote', n).then(() => commit('removeChange', { type: 'add', payload: n })))
            })
            state.changes.delete.forEach(n => {
                promises.push(dispatch('deleteNote', n).then(() => commit('removeChange', { type: 'delete', payload: n })))
            })

            return Promise.all(promises)
        }
    },
    strict: true
})

export default store