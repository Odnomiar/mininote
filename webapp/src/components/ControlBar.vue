<template>
  <div>
    <core-modal v-if="showDiscardModal" @ok="reset" @discard="showDiscardModal = false;">
      <template v-slot:header>Save changes?</template>
      <template v-slot:main>
        <span>You're about to discard all recent changes to your notebook. Are you sure you want to proceed?</span>
      </template>
    </core-modal>

    <core-modal v-if="showSettingsModal" @ok="updateNotebook" @discard="showSettingsModal = false;">
      <template v-slot:header>Notebook Settings</template>
      <template v-slot:main>
        <div class="flex flex-col space-y-4">
          <div class="flex">
            <div class="w-1/4">
              <label for="newNameInput">Name:</label>
            </div>
            <div class="w-3/4">
              <input id="newNameInput" type="text" placeholder="Enter new name for this notebook." v-model="inputs.newName" class="w-full bg-gray-100 focus:shadow-outline" style="border: 1px solid #CBD5E0;" />
            </div>
          </div>
          <div class="flex">
            <div class="w-1/4">
              <label for="newPasswordInput">Password:</label>
            </div>
            <div class="w-3/4">
               <input id="newPasswordInput" type="password" placeholder="Enter new password for this notebook." v-model="inputs.newPassword" class="w-full bg-gray-100 focus:shadow-outline" style="border: 1px solid #CBD5E0;" />
            </div>
          </div>
        </div>
      </template>
      <template v-slot:ok-btn>Save</template>
    </core-modal>

    <div class="flex justify-between">
      <div class="w-1/4 text-2xl text-gray-400">
        <span>âœŽ Mini_Note</span>
        <span class="ml-1 text-xs text-green-600">BETA</span>
      </div>
      <div class="flex justify-center w-1/2" id="notebook-chooser">
        <button
          type="button"
          class="px-4 rounded-r-none btn-danger hover:bg-red-800"
          @click="close"
          :disabled="dirty"
        >&#x2573;</button>
        <input
          type="text"
          class="placeholder-gray-700 rounded-l-none rounded-r-none"
          placeholder="Open or create notebook ..."
          ref="refNotebookInput"
          v-model="inputs.name"
          v-if="!state.opening && !state.creating"
          :disabled="state.loaded"
          @keyup.enter="tryNotebook"
          autofocus
        />
        <input
          type="password"
          class="placeholder-gray-700 rounded-l-none rounded-r-none"
          placeholder="Choose a password for the new notebook..."
          ref="refCreatePasswordInput"
          v-model="inputs.password"
          v-if="state.creating"
          @keyup.enter="createNotebook"
        />
        <input
          type="password"
          class="placeholder-gray-700 rounded-l-none rounded-r-none"
          placeholder="Enter password ..."
          ref="refOpenPasswordInput"
          v-model="inputs.password"
          v-if="state.opening"
          @keyup.enter="openNotebook"
        />
        <button
          type="button"
          class="rounded-l-none btn-primary hover:bg-green-700"
          :disabled="!inputs.name || state.loaded"
          @click="tryNotebook"
          v-if="!state.opening && !state.creating"
        >Open</button>
        <button
          type="button"
          class="rounded-l-none btn-primary hover:bg-green-700"
          :disabled="!inputs.password"
          @click="openNotebook"
          v-if="state.opening"
        >Open</button>
        <button
          type="button"
          class="rounded-l-none btn-primary hover:bg-green-700"
          :disabled="!inputs.password"
          @click="createNotebook"
          v-if="state.creating"
        >Create</button>
      </div>
      <div class="flex justify-end w-1/4 space-x-1">
        <button
          class="px-4 btn-primary hover:bg-green-700"
          @click="saveNotes"
          v-if="dirty"
        >&#x1f4be;</button>
        <button class="px-4 btn-primary hover:bg-green-700" @click="tryReset" v-if="dirty">&#x21ba;</button>
        <button
          class="px-4 btn-primary hover:bg-green-700"
          v-if="state.loaded"
          @click="showSettingsModal = true"
        >&#x2699;</button>
      </div>
    </div>
  </div>
</template>

<script>
import api from "../api";
import { md5 } from "../lib/md5";
import { mapState, mapGetters, mapActions } from "vuex";
import CoreModal from "./core/CoreModal";

export default {
  name: "control-bar",
  components: {
    CoreModal
  },
  data() {
    return {
      inputs: {
        name: "",
        password: "",
        newName: "",
        newPassword: ""
      },
      state: {
        opening: false,
        creating: false,
        loaded: false
      },
      showDiscardModal: false,
      showSettingsModal: false
    };
  },
  computed: {
    ...mapState({
      notebook: state => state.notebook,
      loadNotebookId: state => state.loadNotebookId
    }),
    ...mapGetters(["dirty"])
  },
  methods: {
    ...mapActions(["listNotebooks"]),
    handleError: function(err) {
      this.$emit("alert", err.message);
      this.reset();
    },
    tryReset: function() {
      if (!this.dirty) this.reset();
      else this.showDiscardModal = true;
    },
    reset: function() {
      this.$store.commit("revertChanges");
      this.showDiscardModal = false;
    },
    close: function() {
      this.inputs = {
        name: "",
        password: "",
        newName: "",
        newPassword: ""
      };
      this.state = {
        opening: false,
        creating: false,
        loaded: false
      };
      this.$store.commit("reset");
      this.listNotebooks();
      setTimeout(() => this.$refs.refNotebookInput.focus(), 0);
    },
    tryNotebook: function() {
      let vm = this;
      if (!this.inputs.name) return;

      api
        .exists(this.inputs.name.toLowerCase())
        .then(exists => {
          if (exists) {
            vm.state.opening = true;
            setTimeout(() => vm.$refs.refOpenPasswordInput.focus(), 0);
          } else {
            vm.state.creating = true;
            setTimeout(() => vm.$refs.refCreatePasswordInput.focus(), 0);
          }
        })
        .catch(() => vm.handleError(null));
    },
    openNotebook: function() {
      let vm = this;
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store
        .dispatch("loadNotebook", {
          id: this.inputs.name.toLowerCase(),
          password: md5(this.inputs.password)
        })
        .then(() => {
          vm.state.opening = false;
          vm.state.loaded = true;
          vm.$store.commit("selectFirst");

          this.inputs.newName = this.inputs.name;
          this.inputs.newPassword = this.inputs.password;
        })
        .catch(vm.handleError);
    },
    createNotebook: function() {
      let vm = this;
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store
        .dispatch("createNotebook", {
          id: this.inputs.name.toLowerCase(),
          password: md5(this.inputs.password)
        })
        .then(() => {
          vm.state.creating = false;
          vm.state.loaded = true;
        })
        .catch(vm.handleError);
    },
    saveNotes: function() {
      let vm = this;
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store
        .dispatch("applyChanges")
        .then(() => this.$emit("alert", "Saved changes.", "success"))
        .catch(vm.handleError);
    },
    updateNotebook: function() {
      let vm = this;
      if (!this.inputs.name || !this.inputs.password) return;

      this.$store
        .dispatch("updateNotebook", {
          id: this.inputs.newName.toLowerCase(),
          password: md5(this.inputs.newPassword)
        })
        .then(() => {
          (this.inputs.name = this.inputs.newName),
            (this.inputs.password = this.inputs.newPassword);
          this.$emit("alert", "Notebook updated.", "success");
          this.showSettingsModal = false;
        })
        .catch(vm.handleError);
    }
  },
  watch: {
    loadNotebookId: function(newId, oldId) {
      if (newId !== "" && newId !== oldId) {
        this.inputs.name = newId;
      }
    }
  }
};
</script>

<style scoped>
</style>
