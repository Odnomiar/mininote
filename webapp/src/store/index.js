import Vuex from 'vuex'

import api from '../api'

import { version } from '../../package.json'

/* eslint-disable no-case-declarations */

const emptyState = {
  notebook: {},
  notebooks: [],
  changes: {
    add: [],
    delete: [],
    update: {},
  },
  selectedNoteId: -1,
  loadNotebookId: '',
  version,
}

/// GETTERS ///
function loaded(state) {
  return !!state.notebook.notes
}

function dirty(state) {
  return !!(
    state.changes.add.length ||
    state.changes.delete.length ||
    Object.keys(state.changes.update).length
  )
}

function selectedNote(state) {
  return resolveNotes(state).find((n) => n.id === state.selectedNoteId)
}

function noteById(state) {
  return (id) => resolveNotes(state).find((n) => n.id === id)
}

function resolveNotes(state) {
  let notes = []
  notes = notes.concat(state.notebook.notes || [])
  notes = notes.concat(state.changes.add)
  notes = notes.filter(
    (n1) => !state.changes.delete.some((n2) => n1.id === n2.id),
  )
  return notes
}

function currentContent(state) {
  let id = state.selectedNoteId
  if (id < 0) return ''
  if (id in state.changes.update) return state.changes.update[id].content
  const note = selectedNote(state)
  return note ? note.content : ''
}

function currentTitle(state) {
  let id = state.selectedNoteId
  if (id < 0) return ''
  if (id in state.changes.update) return state.changes.update[id].title
  const note = selectedNote(state)
  return note ? note.title : ''
}

/// MUTATIONS ///
function selectNote(state, id) {
  if (id < 0) state.selectedNoteId = -1
  let notes = resolveNotes(state)
  if (notes.some((n) => n.id === id)) {
    state.selectedNoteId = id
  }
}

function selectFirst(state) {
  let notes = resolveNotes(state)
  let id = notes.length ? notes[0].id : -1
  state.selectedNoteId = id
}

function updateUrl(state) {
  location.hash = state.notebook.id ? `#/notebook/${state.notebook.id}` : ''
}

/// STORE ///
const store = new Vuex.Store({
  state: JSON.parse(JSON.stringify(emptyState)),
  getters: {
    selectedNote,
    dirty,
    notes: resolveNotes,
    currentContent,
    currentTitle,
    loaded,
    noteById,
  },
  mutations: {
    setNotebooks(state, notebookList) {
      state.notebooks = notebookList
    },
    setNotebook(state, { id, password }) {
      state.notebook.id = id
      state.notebook.password = password
      updateUrl(state)
    },
    setNotes(state, notes) {
      state.notebook.notes = notes
    },
    addNote(state, note) {
      state.notebook.notes = state.notebook.notes.concat([note])
      if (state.selectedNoteId === note.oldId) selectNote(state, note.id)
    },
    removeNote(state, note) {
      state.notebook.notes = state.notebook.notes.filter(
        (n) => n.id !== note.id,
      )
      if (state.selectedNoteId === note.id && state.notebook.notes.length) {
        state.selectedNoteId = state.notes[0].id
      }
    },
    updateNote(state, note) {
      let notes = resolveNotes(state)
      let found = notes.find((n) => n.id === note.id)
      for (let prop in found) {
        found[prop] = note[prop]
      }
    },
    selectNote,
    selectFirst,
    setLoadNotebook(state, id) {
      state.loadNotebookId = id
    },
    addChange(state, { type, payload }) {
      switch (type) {
        case 'add':
        case 'delete':
          state.changes[type] = state.changes[type].concat([payload])
          break
        case 'update':
          const hasChange = payload.id in state.changes.update
          const originalNote = (state.notebook.notes || []).find(
            (n) => n.id === payload.id,
          )

          if (
            originalNote &&
            originalNote.content === payload.content &&
            hasChange
          ) {
            delete state.changes.update[payload.id]
          } else {
            state.changes.update[payload.id] = payload
          }
          break
      }
    },
    removeChange(state, { type, payload }) {
      switch (type) {
        case 'add':
        case 'delete':
          state.changes[type] = state.changes[type].filter(
            (n) => n.id !== payload.id,
          )
          break
        case 'update':
          delete state.changes.update[payload.id]
      }
    },
    revertChanges(state) {
      state.changes.add = []
      state.changes.delete = []

      Object.keys(state.changes.update).forEach((id) => {
        let note = (state.notebook.notes || []).find(
          (n) => n.id === parseInt(id),
        )

        if (!note) return

        let content = note.content
        note.content = ''
        note.content = content
      })

      state.changes.update = {}

      if (!resolveNotes(state).length) {
        selectFirst(state) // select none, i.e. -1
      }
    },
    reset(state) {
      for (let prop in emptyState) {
        state[prop] = JSON.parse(JSON.stringify(emptyState[prop]))
      }
      updateUrl(state)
    },
  },
  actions: {
    listNotebooks({ commit }) {
      return api.list().then((data) => commit('setNotebooks', data))
    },
    loadNotebook({ commit }, { id, password }) {
      return api
        .getNotes(id, password)
        .then((data) => commit('setNotes', data))
        .then(() => commit('setNotebook', { id, password }))
    },
    createNotebook({ commit }, { id, password }) {
      return api
        .create(id, password)
        .then((data) => commit('setNotes', data.notes))
        .then(() => commit('setNotebook', { id, password }))
    },
    updateNotebook({ state, commit }, notebook) {
      let updateNotebook = JSON.parse(JSON.stringify(notebook))
      if (notebook.id === state.notebook.id) delete updateNotebook.id
      return api
        .update(state.notebook.id, state.notebook.password, updateNotebook)
        .then(() => commit('setNotebook', notebook))
    },
    createNote({ state, commit }, note) {
      return api
        .addNote(state.notebook.id, state.notebook.password, note)
        .then((data) =>
          commit('addNote', { ...note, id: data.id, oldId: note.id }),
        )
    },
    updateNote({ state, commit }, note) {
      return api
        .updateNote(state.notebook.id, state.notebook.password, note)
        .then(() => commit('updateNote', note))
    },
    deleteNote({ state, commit }, note) {
      return api
        .deleteNote(state.notebook.id, state.notebook.password, note)
        .then(() => commit('removeNote', note))
    },
    applyChanges({ state, commit, dispatch }) {
      let promises = []

      Object.values(state.changes.update).forEach((n) => {
        let newNote = state.changes.add.find((n1) => n1.id === n.id)
        let deletedNote = state.changes.delete.find((n1) => n1.id === n.id)

        if (newNote) {
          commit('updateNote', n)
          commit('removeChange', { type: 'update', payload: n })
        } else if (deletedNote)
          commit('removeChange', { type: 'update', payload: n })
        else
          promises.push(
            dispatch('updateNote', n).then(() =>
              commit('removeChange', { type: 'update', payload: n }),
            ),
          )
      })

      state.changes.add.forEach((n) => {
        if (state.changes.delete.some((n1) => n1.id === n.id)) {
          commit('removeChange', { type: 'add', payload: n })
          commit('removeChange', { type: 'delete', payload: n })
        } else
          promises.push(
            dispatch('createNote', n).then(() =>
              commit('removeChange', { type: 'add', payload: n }),
            ),
          )
      })
      state.changes.delete.forEach((n) => {
        promises.push(
          dispatch('deleteNote', n).then(() =>
            commit('removeChange', { type: 'delete', payload: n }),
          ),
        )
      })

      return Promise.all(promises)
    },
  },
  strict: true,
})

export function useStore() {
  return store
}
