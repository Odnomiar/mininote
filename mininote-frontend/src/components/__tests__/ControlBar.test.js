import { createLocalVue, shallowMount } from '@vue/test-utils'
import BootstrapVue from 'bootstrap-vue'
import ControlBar from '@/components/ControlBar'
import store from '../../store'

const localVue = createLocalVue()

localVue.use(BootstrapVue)

describe('ControlBar.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(ControlBar, {
      localVue,
      propsData: {
        hasChanges: false,
        notes: ['note one', 'note two'],
      },
      store
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
