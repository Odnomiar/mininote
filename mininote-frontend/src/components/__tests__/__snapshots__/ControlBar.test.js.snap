// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlBar.vue matches snapshot 1`] = `
<div notes="note one,note two">
  <bmodal-stub title="Save changes?" titletag="h5" size="md" buttonsize="" headercloselabel="Close" canceltitle="Cancel" oktitle="OK" cancelvariant="secondary" okvariant="primary" id="discardModal">You're about to discard all recent changes to your notebook. Are you sure you want to proceed?</bmodal-stub>
  <bmodal-stub title="Notebook Settings" titletag="h5" size="lg" buttonsize="" headercloselabel="Close" canceltitle="Cancel" oktitle="Save" cancelvariant="secondary" okvariant="primary" id="settingsModal">
    <bcontainer-stub tag="div" fluid="true">
      <brow-stub tag="div">
        <bcol-stub sm="3" tag="div"><label for="newNameInput">Name:</label></bcol-stub>
        <bcol-stub tag="div">
          <bforminput-stub value="" type="text" placeholder="Enter new name for this notebook." id="newNameInput"></bforminput-stub>
        </bcol-stub>
      </brow-stub>
      <brow-stub tag="div">
        <bcol-stub sm="3" tag="div"><label for="newPasswordInput">Password:</label></bcol-stub>
        <bcol-stub tag="div">
          <bforminput-stub value="" type="password" placeholder="Enter new password for this notebook." id="newPasswordInput"></bforminput-stub>
        </bcol-stub>
      </brow-stub>
    </bcontainer-stub>
  </bmodal-stub>
  <div class="container-fluid">
    <div class="row">
      <div class="col-4"></div>
      <div class="col-4">
        <div id="notebook-chooser" class="input-group">
          <div class="input-group-btn"><button type="button" class="btn btn-danger">â•³</button></div> <input type="text" placeholder="Open or create notebook ..." autofocus="autofocus" class="form-control">
          <!---->
          <!---->
          <div class="input-group-btn"><button type="button" disabled="disabled" class="btn btn-primary">Open</button>
            <!---->
            <!---->
          </div>
        </div>
      </div>
      <div class="col-2"></div>
      <div class="col-2 action-buttons-container">
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;
